### DISCUSSION TITLE
Spifffs programming

### DISCUSSION THREAD

**Discussion by Mair** [Mair/esp32-course#64] [Course Qeustions] [answered]
from David
Mair,

Are you up on the versions of esp extensions in VScode?
I thought now if you programmed a device and it was using a spiffs partition that it programmed the partition automatically?
Or, do I need to program the partition separately and program the code?
I have loaded up my logger code that I completed a while ago but when I go to run in device I get “Failed to open for writing” when I attempt to create the file.
My suspicion is that it has not programmed the partition

**Comment by Mair** [comment, marked as answer]
Hi David,

For SPIFFS to be programmed into flash when flashing the chip, you need to add spiffs_create_partition_image() into your CMakelists.txt. 
Please see [https://learnesp32.com/videos/8/6_spiffs%20-%20upload%20with%20flash](https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Flearnesp32.com%2Fvideos%2F8%2F6_spiffs%2520-%2520upload%2520with%2520flash&data=04%7C01%7C%7C505209e6be994f0d6e3208da15ff7246%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637846484123588217%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=%2B134IYzMOBHSstnoH%2BYzhUIzig%2BspkqtDnYOcsNqW8g%3D&reserved=0) time=3:25

**Reply by orbitcoms** [reply]
Hi Mair,
I do have the FLASH_IN_PROJECT line in the makelist ok. and have not changed these since the software was last running ok.
It is difficult to see what the compiler is doing because the build output is not appearing in the VS code output window, it remains blank and I just see "Build Successful" at end. I though there must be a "silent build" option somewhere I can turn off in vscode but cannot find it.

**Reply by orbitcoms** [reply]
Still same problem. I have attached the build output if it helps.

![build2](https://user-images.githubusercontent.com/41096656/161662360-49759385-d41d-40e3-b2d5-be9f493a4286.JPG)
![build1](https://user-images.githubusercontent.com/41096656/161662367-33b88781-b1ff-4031-87f6-fff8304b36ec.JPG)

**Reply by Mair** [reply]
Hi @David. You can do a diagnosis as follows.

 When the chip starts up it prints out the partition table
![image](https://user-images.githubusercontent.com/2477839/161725697-09edb307-6cc5-4283-8a2b-e816c781e5f9.png)
Notice the start address of my SPIFFS partition
Now look before then when it was flashing the chip and notice it was writing to the address
![image](https://user-images.githubusercontent.com/2477839/161726105-30bf7610-4ea2-4a83-a96e-d060658d115d.png)

You can also add the following at the beginning of your code

```c
 esp_vfs_spiffs_conf_t esp_vfs_spiffs_conf = {
      .base_path = "/spiffs",
      .partition_label = NULL,
      .max_files = 5,
      .format_if_mount_failed = true};
  esp_vfs_spiffs_register(&esp_vfs_spiffs_conf);

  size_t total = 0;
  size_t used = 0;
  esp_spiffs_info(NULL, &total, &used);

  ESP_LOGI("SPIFFS", "total %d, used %d", total, used);
  esp_vfs_spiffs_unregister(NULL);
```
This will print out the total SPIFFS and the used. Let me know what you find. The error might be code related and the SPIFFS data is present

**Comment by Mair** [comment]
Interestingly I think I had the same issue which didn't happen if I used the command line.
Try opening an idf terminal and use idf.py flash monitor it should show the output and flash to spiffs

Let me know if this resolves it.

Regards,
Mair

**Reply by orbitcoms** [reply]
Thanks MAir,

Having so much trouble with VSCODE. There were errors for cmake commands not being found and the output kept defaulting to hidden so I could not see what was being compiled.

My app has 2 spiffs partitions.

One used for a log and other for html

While I am waiting to try again, should I de-register spiffs every time after use or can I register it and leave it running to be able to simply add to my file as required?

**Reply by Mair** [reply]
the latter is fine, Just be aware the code I presented assumes 1 partition. You will need to change the labels for it to work
```c
 esp_vfs_spiffs_conf_t esp_vfs_spiffs_conf = {
      .base_path = "/spiffs",
      .partition_label = "<label>"
      .max_files = 5,
      .format_if_mount_failed = true};
  esp_vfs_spiffs_register(&esp_vfs_spiffs_conf);

  size_t total = 0;
  size_t used = 0;
  esp_spiffs_info( "<label>", &total, &used);

  ESP_LOGI("SPIFFS", "total %d, used %d", total, used);
  esp_vfs_spiffs_unregister( "<label>";

```

**Reply by orbitcoms** [reply]
Even with new install I cannot even get it to compile now. Some issue with register command?
![no_compile](https://user-images.githubusercontent.com/41096656/161733988-091dc449-e2f9-41c5-b90c-422bab16f22c.JPG)

**Comment by Mair** [comment]
No. This is an error with your CMakeLists.txt.
You have a syntax error or you have registered a file in there "connect.c" which it cand find

**Reply by orbitcoms** [reply]
I'm perplexed. Had so much trouble all day yet thise code was used last time no problm and not altered since itr was used.
I can see connect.c in the folder ok?

![no_compile](https://user-images.githubusercontent.com/41096656/161735413-b1820936-bcba-4650-9ce1-ee29b4677e50.JPG)

**Reply by Mair** [reply]
that looks OK. not sure whats happening there. maybe try a full clean and then rebuild. 
Can you build any projects? try building a hello world and see if its your environment or project related.

**Reply by orbitcoms** [reply]
I have cleaned and rebuilt several times.
Other projects were ok and a new test project with just one partition worked. 

No idea what is happening here. Having to reinstall esp-idf again in VScode because it could not find the toools folder files now.

**Comment by Mair** [comment]
Its not going to be an issue with vscode. It doesn't do anything to your compilation.
If you can build other projects then it's not your IDF installation. Its more than likely a fault in the project somewhere. I would need to have a closer look to debug it.

**Reply by orbitcoms** [reply]
I think I have a bit of progress.

I now at least see it is trying to organise the spiffs partitions and complaining I have not used correct base address
![Uploading no_compile.JPG…]()

**Reply by orbitcoms** [reply]
Now it compiles but get "Failed to initialize Spiffs (ESP_ERR_INVALID_ARG)  //There must be an error or wrong folder pointed to

I have the 2 folders to program into flash above the main folder labelled
html
log

The partitions are set up as 
log_storage,  data,  spiffs, ,        2M,
html_storage, data, spiffs ,,   100K

In cmakelists I do this
spiffs_create_partition_image(html_storage ../html FLASH_IN_PROJECT)
spiffs_create_partition_image(log_storage ../log FLASH_IN_PROJECT)

The code that uses log
  esp_vfs_spiffs_conf_t log_config = 
  {
    .base_path = "spiffs",
    .partition_label = "log_storage",
    .max_files = 5,
    .format_if_mount_failed = true,
  };
  esp_err_t ret = esp_vfs_spiffs_register(&log_config);

then the FILE commands attempt to open and then create file if not found.
#define log_file "log.txt"   ?is my filename correct or need to be preceeded with something?

and code that uses html
 esp_vfs_spiffs_conf_t html_config = {
        .base_path = "spiffs",
        .partition_label = "html_storage",
        .max_files = 5,
        .format_if_mount_failed = true};
    esp_vfs_spiffs_register(&html_config);

**Reply by orbitcoms** [reply]
I can see the flash is being written ok to the partitions at 0011.... (log storage) and 0021...(html_storage)

So, must be issue in spiffs init, are my base path and label correct ?

esp_vfs_spiffs_conf_t log_config = 
  {
    .base_path = "spiffs",
    .partition_label = "log_storage",
    .max_files = 5,
    .format_if_mount_failed = true,
  };

  esp_err_t ret = esp_vfs_spiffs_register(&log_config);

It is failing in the "else" catch below... with an INVALID ARG response

  if(ret != ESP_OK)
  {
      if(ret == ESP_FAIL)
      {
          ESP_LOGE(TAG,"Failed to mount or format file system");
      }
      else if(ret == ESP_ERR_NOT_FOUND)
      {
          ESP_LOGE(TAG,"Failed to find spiffs partition");
      }
      else
      {
          ESP_LOGE(TAG,"Failed to initialize Spiffs (%s)",esp_err_to_name(ret));
      }
      return;
  }

**Reply by orbitcoms** [reply]
Thanks for all your help

Problem was .base_path, it had to be "/log"

**Reply by Mair** [reply]
@orbitcoms glad to hear it 👍
